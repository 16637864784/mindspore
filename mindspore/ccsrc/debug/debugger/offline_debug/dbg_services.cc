/**
 * Copyright 2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "debugger/offline_debug/dbg_services.h"

#include <algorithm>
#include <chrono>

DbgServices::DbgServices(bool verbose) {
  DbgLogger::verbose = verbose;
  char *dbg_log_path = getenv("OFFLINE_DBG_LOG");
  if (dbg_log_path != NULL) {
    DbgLogger::verbose = true;
  }
  debug_services = new DebugServices();
}

DbgServices::DbgServices(const DbgServices &other) {
  MS_LOG(INFO) << "cpp DbgServices object is created via copy";
  debug_services = new DebugServices(*other.debug_services);
}

DbgServices &DbgServices::operator=(const DbgServices &other) {
  MS_LOG(INFO) << "cpp DbgServices object is being assigned a different state";
  if (this != &other) {
    delete debug_services;
    debug_services = new DebugServices(*other.debug_services);
  }
  return *this;
}

DbgServices::~DbgServices() {
  MS_LOG(INFO) << "cpp DbgServices object is deleted";
  delete debug_services;
}

std::string DbgServices::GetVersion() {
  MS_LOG(INFO) << "get version is called";
  return "1.2.0";
}

int32_t DbgServices::Initialize(std::string net_name, std::string dump_folder_path, bool is_sync_mode) {
  MS_LOG(INFO) << "cpp DbgServices initialize network name " << net_name;
  MS_LOG(INFO) << "cpp DbgServices initialize dump folder path " << dump_folder_path;
  MS_LOG(INFO) << "cpp DbgServices initialize sync mode " << is_sync_mode;
  if (debug_services == nullptr) {
    MS_LOG(EXCEPTION) << "Debugger services initialize failed as occur null pointer error,"
                      << "may be due to memory allocation failure, check as: top";
  }
  debug_services->SetNetName(net_name);
  debug_services->SetDumpDir(dump_folder_path);
  debug_services->SetSyncMode(is_sync_mode);
  return 0;
}

int32_t DbgServices::AddWatchpoint(
  unsigned int id, unsigned int watch_condition,
  std::map<std::string, std::map<std::string, std::variant<bool, std::vector<std::string>>>> check_nodes,
  std::vector<parameter_t> parameter_list) {
  MS_LOG(INFO) << "cpp start";

  MS_LOG(INFO) << "cpp DbgServices AddWatchpoint id " << id;
  MS_LOG(INFO) << "cpp DbgServices AddWatchpoint watch_condition " << watch_condition;
  for (auto const &node : check_nodes) {
    MS_LOG(INFO) << "cpp DbgServices AddWatchpoint name " << node.first;
    auto attr_map = node.second;

    bool is_parameter = std::get<bool>(attr_map["is_parameter"]);
    MS_LOG(INFO) << "cpp DbgServices AddWatchpoint is_parameter " << is_parameter;

    // std::vector<uint32_t> device_id = std::get<std::vector<uint32_t>>(attr_map["device_id"]);
    std::vector<std::string> device_id_str = std::get<std::vector<std::string>>(attr_map["device_id"]);
    std::vector<std::uint32_t> device_id;
    std::transform(device_id_str.begin(), device_id_str.end(), std::back_inserter(device_id),
                   [](std::string &id_str) -> std::uint32_t { return static_cast<uint32_t>(std::stoul(id_str)); });
    MS_LOG(INFO) << "cpp DbgServices AddWatchpoint device_id ";
    for (auto const &i : device_id) {
      MS_LOG(INFO) << i << " ";
    }

    // std::vector<uint32_t> root_graph_id = std::get<std::vector<uint32_t>>(attr_map["root_graph_id"]);
    std::vector<std::string> root_graph_id_str = std::get<std::vector<std::string>>(attr_map["root_graph_id"]);
    std::vector<std::uint32_t> root_graph_id;
    std::transform(
      root_graph_id_str.begin(), root_graph_id_str.end(), std::back_inserter(root_graph_id),
      [](std::string &graph_str) -> std::uint32_t { return static_cast<uint32_t>(std::stoul(graph_str)); });
    MS_LOG(INFO) << "cpp DbgServices AddWatchpoint root_graph_id";
    for (auto const &j : root_graph_id) {
      MS_LOG(INFO) << j << " ";
    }
  }

  for (auto const &parameter : parameter_list) {
    MS_LOG(INFO) << "cpp DbgServices AddWatchpoint parameter name " << parameter.name;
    MS_LOG(INFO) << "cpp DbgServices AddWatchpoint parameter disabled " << parameter.disabled;
    MS_LOG(INFO) << "cpp DbgServices AddWatchpoint parameter value " << parameter.value;
    MS_LOG(INFO) << "cpp DbgServices AddWatchpoint parameter hit " << parameter.hit;
    MS_LOG(INFO) << "cpp DbgServices AddWatchpoint parameter actual_value " << parameter.actual_value;
  }

  std::vector<std::tuple<std::string, bool>> check_node_list;
  std::vector<std::tuple<std::string, std::vector<uint32_t>>> check_node_device_list;
  std::vector<std::tuple<std::string, std::vector<uint32_t>>> check_node_graph_list;
  std::vector<DebugServices::parameter_t> parameter_list_backend;

  std::transform(check_nodes.begin(), check_nodes.end(), std::back_inserter(check_node_list),
                 [](auto &node) -> std::tuple<std::string, bool> {
                   auto attr_map = node.second;
                   return std::make_tuple(node.first, std::get<bool>(attr_map["is_parameter"]));
                 });

  std::transform(check_nodes.begin(), check_nodes.end(), std::back_inserter(check_node_device_list),
                 [](auto &node) -> std::tuple<std::string, std::vector<uint32_t>> {
                   auto attr_map = node.second;
                   std::vector<std::string> device_id_str = std::get<std::vector<std::string>>(attr_map["device_id"]);
                   std::vector<std::uint32_t> device_id;
                   std::transform(
                     device_id_str.begin(), device_id_str.end(), std::back_inserter(device_id),
                     [](std::string &id_str) -> std::uint32_t { return static_cast<uint32_t>(std::stoul(id_str)); });
                   return std::make_tuple(node.first, device_id);
                 });

  std::transform(
    check_nodes.begin(), check_nodes.end(), std::back_inserter(check_node_graph_list),
    [](auto &node) -> std::tuple<std::string, std::vector<uint32_t>> {
      auto attr_map = node.second;
      std::vector<std::string> root_graph_id_str = std::get<std::vector<std::string>>(attr_map["root_graph_id"]);
      std::vector<std::uint32_t> root_graph_id;
      std::transform(
        root_graph_id_str.begin(), root_graph_id_str.end(), std::back_inserter(root_graph_id),
        [](std::string &graph_str) -> std::uint32_t { return static_cast<uint32_t>(std::stoul(graph_str)); });
      return std::make_tuple(node.first, root_graph_id);
    });

  std::transform(
    parameter_list.begin(), parameter_list.end(), std::back_inserter(parameter_list_backend),
    [](const parameter_t &parameter) -> DebugServices::parameter_t {
      return DebugServices::parameter_t{parameter.name, parameter.disabled, parameter.value, parameter.hit};
    });

  debug_services->AddWatchpoint(id, watch_condition, 0, check_node_list, parameter_list_backend,
                                &check_node_device_list, &check_node_graph_list);
  MS_LOG(INFO) << "cpp end";
  return 0;
}

int32_t DbgServices::RemoveWatchpoint(unsigned int id) {
  MS_LOG(INFO) << "cpp DbgServices RemoveWatchpoint id " << id;
  debug_services->RemoveWatchpoint(id);
  return 0;
}

std::vector<watchpoint_hit_t> DbgServices::CheckWatchpoints(unsigned int iteration) {
  MS_LOG(INFO) << "cpp DbgServices CheckWatchpoint iteration " << iteration;

  std::vector<std::string> name;
  std::vector<std::string> slot;
  std::vector<int> condition;
  std::vector<unsigned int> watchpoint_id;
  std::vector<std::string> overflow_ops;
  std::vector<std::vector<DebugServices::parameter_t>> parameters;
  std::vector<int32_t> error_codes;
  std::vector<unsigned int> device_id;
  std::vector<unsigned int> root_graph_id;
  // #ifdef ENABLE_D
  //  overflow_ops = CheckOpOverflow();
  // #endif

  std::vector<std::shared_ptr<TensorData>> tensor_list;
  std::vector<std::string> file_paths;

  tensor_list = debug_services->ReadNeededDumpedTensors(iteration, &file_paths);

  debug_services->CheckWatchpoints(&name, &slot, &condition, &watchpoint_id, &parameters, &error_codes, overflow_ops,
                                   file_paths, &tensor_list, false, true, true, &device_id, &root_graph_id);

  std::vector<watchpoint_hit_t> hits;
  for (unsigned int i = 0; i < name.size(); i++) {
    std::vector<DebugServices::parameter_t> &parameter = parameters[i];
    std::vector<parameter_t> api_parameter_vector;
    for (const auto &p : parameter) {
      parameter_t api_parameter(p.name, p.disabled, p.value, p.hit, p.actual_value);
      api_parameter_vector.push_back(api_parameter);
    }
    watchpoint_hit_t hit(name[i], std::stoi(slot[i]), condition[i], watchpoint_id[i], api_parameter_vector,
                         error_codes[i], device_id[i], root_graph_id[i]);

    MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t name " << hit.name;
    MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t slot " << hit.slot;
    MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t watchpoint_id " << hit.watchpoint_id;
    MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t error_code " << hit.error_code;
    MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t device_id " << hit.device_id;
    MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t root_graph_id " << hit.root_graph_id;

    for (auto const &parameter_i : api_parameter_vector) {
      MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t parameter name " << parameter_i.name;
      MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t parameter disabled " << parameter_i.disabled;
      MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t parameter value " << parameter_i.value;
      MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t parameter hit " << parameter_i.hit;
      MS_LOG(INFO) << "cpp DbgServices watchpoint_hit_t parameter actual_value " << parameter_i.actual_value;
    }

    hits.push_back(hit);
  }
  return hits;
}

std::string GetTensorFullName(tensor_info_t info) {
  std::string node_name = info.node_name;
  if (info.is_parameter) {
    // scopes in node name are separated by '/'
    // use the name without scope if truncate is true
    auto found = node_name.find_last_of("/");
    if (found != std::string::npos) {
      node_name = node_name.substr(found + 1);
    }
  }
  return node_name + ":" + std::to_string(info.slot);
}

unsigned int GetTensorDeviceId(tensor_info_t info) { return info.device_id; }

unsigned int GetTensorRootGraphId(tensor_info_t info) { return info.root_graph_id; }

unsigned int GetTensorIteration(tensor_info_t info) { return info.iteration; }

unsigned int GetTensorSlot(tensor_info_t info) { return info.slot; }

std::vector<tensor_data_t> DbgServices::ReadTensors(std::vector<tensor_info_t> info) {
  for (auto i : info) {
    MS_LOG(INFO) << "cpp DbgServices ReadTensor info name " << i.node_name << ", slot " << i.slot << ", iteration "
                 << i.iteration << ", device_id " << i.device_id << ", root_graph_id " << i.root_graph_id;
  }
  std::vector<std::string> backend_name;
  std::vector<unsigned int> device_id;
  std::vector<unsigned int> root_graph_id;
  std::vector<unsigned int> iteration;
  std::vector<size_t> slot;
  std::vector<std::shared_ptr<TensorData>> result_list;
  std::vector<tensor_data_t> tensors_read;

  std::transform(info.begin(), info.end(), std::back_inserter(backend_name), GetTensorFullName);
  std::transform(info.begin(), info.end(), std::back_inserter(slot), GetTensorSlot);
  std::transform(info.begin(), info.end(), std::back_inserter(device_id), GetTensorDeviceId);
  std::transform(info.begin(), info.end(), std::back_inserter(root_graph_id), GetTensorRootGraphId);
  std::transform(info.begin(), info.end(), std::back_inserter(iteration), GetTensorIteration);

  MS_LOG(INFO) << "cpp before";
  std::vector<std::string> file_paths;
  auto t1 = std::chrono::high_resolution_clock::now();
  debug_services->ConvertReadTensors(backend_name, slot, device_id, iteration, root_graph_id, &file_paths);
  debug_services->ReadDumpedTensor(backend_name, slot, device_id, iteration, root_graph_id, file_paths, &result_list);
  auto t2 = std::chrono::high_resolution_clock::now();
  /* Getting number of milliseconds as a double. */
  std::chrono::duration<double, std::milli> ms_double = t2 - t1;

  MS_LOG(INFO) << "ReadTensors Took: " << ms_double.count() / 1000 << "s";
  MS_LOG(INFO) << "cpp after";

  for (auto result : result_list) {
    tensor_data_t tensor_data_item(result->GetDataPtr(), result->GetByteSize(), result->GetType(), result->GetShape());
    tensors_read.push_back(tensor_data_item);
  }
  MS_LOG(INFO) << "cpp end";
  return tensors_read;
}
