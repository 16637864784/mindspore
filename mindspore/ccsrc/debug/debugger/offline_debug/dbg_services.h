/**
 * Copyright 2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#ifndef DEBUG_DBG_SERVICES_H_
#define DEBUG_DBG_SERVICES_H_

#include <vector>
#include <string>
#include <map>
#include <memory>
#include <tuple>
#include <iostream>
#include <variant>
#include "pybind11/pybind11.h"
#include "pybind11/stl.h"
#include "pybind11/stl_bind.h"

#include "debug/debug_services.h"
namespace py = pybind11;

typedef struct parameter {
  parameter(const std::string &name, bool disabled, double value, bool hit, double actual_value)
      : name(name), disabled(disabled), value(value), hit(hit), actual_value(actual_value) {}
  const std::string get_name() const { return name; }
  const bool get_disabled() const { return disabled; }
  const double get_value() const { return value; }
  const bool get_hit() const { return hit; }
  const double get_actual_value() const { return actual_value; }
  std::string name;
  bool disabled;
  double value;
  bool hit;
  double actual_value;
} parameter_t;

typedef struct watchpoint_hit {
  watchpoint_hit(const std::string &name, uint32_t slot, int condition, uint32_t watchpoint_id,
                 const std::vector<parameter_t> &parameters, int32_t error_code, uint32_t device_id,
                 uint32_t root_graph_id)
      : name(name),
        slot(slot),
        condition(condition),
        watchpoint_id(watchpoint_id),
        parameters(parameters),
        error_code(error_code),
        device_id(device_id),
        root_graph_id(root_graph_id) {}
  const std::string get_name() const { return name; }
  const uint32_t get_slot() const { return slot; }
  const int get_condition() const { return condition; }
  const uint32_t get_watchpoint_id() const { return watchpoint_id; }
  const std::vector<parameter_t> get_parameters() const { return parameters; }
  const int32_t get_error_code() const { return error_code; }
  const uint32_t get_device_id() const { return device_id; }
  const uint32_t get_root_graph_id() const { return root_graph_id; }
  std::string name;
  uint32_t slot;
  int condition;
  uint32_t watchpoint_id;
  std::vector<parameter_t> parameters;
  int32_t error_code;
  uint32_t device_id;
  uint32_t root_graph_id;
} watchpoint_hit_t;

typedef struct tensor_info {
  tensor_info(const std::string &node_name, uint32_t slot, uint32_t iteration, uint32_t device_id,
              uint32_t root_graph_id, bool is_parameter)
      : node_name(node_name),
        slot(slot),
        iteration(iteration),
        device_id(device_id),
        root_graph_id(root_graph_id),
        is_parameter(is_parameter) {}
  const std::string get_node_name() const { return node_name; }
  const uint32_t get_slot() const { return slot; }
  const uint32_t get_iteration() const { return iteration; }
  const uint32_t get_device_id() const { return device_id; }
  const uint32_t get_root_graph_id() const { return root_graph_id; }
  const bool get_is_parameter() const { return is_parameter; }
  std::string node_name;
  uint32_t slot;
  uint32_t iteration;
  uint32_t device_id;
  uint32_t root_graph_id;
  bool is_parameter;
} tensor_info_t;

typedef struct tensor_data {
  tensor_data(char *data_ptr, uint64_t data_size, int dtype, const std::vector<int64_t> &shape)
      : data_size(data_size), dtype(dtype), shape(shape) {
    if (data_ptr != NULL) {
      this->data_ptr = py::bytes(data_ptr, data_size);
    } else {
      this->data_ptr = py::bytes();
    }
  }
  const py::bytes get_data_ptr() const { return data_ptr; }
  const uint64_t get_data_size() const { return data_size; }
  const int get_dtype() const { return dtype; }
  const std::vector<int64_t> &get_shape() const { return shape; }
  py::bytes data_ptr;
  uint64_t data_size;
  int dtype;
  std::vector<int64_t> shape;
} tensor_data_t;

class DbgServices {
 private:
  DebugServices *debug_services;

 public:
  explicit DbgServices(bool verbose = false);

  DbgServices(const DbgServices &other);

  DbgServices &operator=(const DbgServices &other);

  ~DbgServices();

  int32_t Initialize(std::string net_name, std::string dump_folder_path, bool is_sync_mode);

  int32_t AddWatchpoint(
    unsigned int id, unsigned int watch_condition,
    std::map<std::string, std::map<std::string, std::variant<bool, std::vector<std::string>>>> check_nodes,
    std::vector<parameter_t> parameter_list);

  int32_t RemoveWatchpoint(unsigned int id);

  std::vector<watchpoint_hit_t> CheckWatchpoints(unsigned int iteration);

  std::vector<tensor_data_t> ReadTensors(std::vector<tensor_info_t> info);

  std::string GetVersion();
};

#endif  // DEBUG_DBG_SERVICES_H_
