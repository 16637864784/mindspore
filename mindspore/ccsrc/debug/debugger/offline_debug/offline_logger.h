/**
 * Copyright 2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#ifndef OFFLINE_LOGGER_H_
#define OFFLINE_LOGGER_H_

#include <iostream>

#define PATH_MAX 4096

#define MS_LOG(level) MS_LOG_##level

#define MS_LOG_INFO static_cast<void>(0), !(DbgLogger::verbose) ? void(0) : DbgLogger(DbgLoggerLvl::INFO) < std::cout

#define MS_LOG_ERROR MS_LOG_INFO

#define MS_LOG_DEBUG MS_LOG_INFO

#define MS_LOG_WARNING MS_LOG_INFO

#define MS_LOG_EXCEPTION static_cast<void>(0), DbgLogger(DbgLoggerLvl::EXCEPTION) < std::cout

enum DbgLoggerLvl : int { DEBUG = 0, INFO, WARNING, ERROR, EXCEPTION };

class DbgLogger {
 public:
  explicit DbgLogger(DbgLoggerLvl lvl) : lvl_(lvl) {}
  ~DbgLogger() = default;
  void operator<(std::ostream &os) const {
    char *dbg_log_path = std::getenv("OFFLINE_DBG_LOG");
    if (dbg_log_path != nullptr) {
      char abspath[PATH_MAX];
      if (sizeof(dbg_log_path) > PATH_MAX || NULL == realpath(dbg_log_path, abspath)) {
        return;
      }
      FILE *fp = freopen(abspath, "a", stdout);
      if (fp == nullptr) {
        std::cout << "ERROR: DbgLogger could not redirect all stdout to a file";
      }
    }
    os << std::endl;
    if (lvl_ == DbgLoggerLvl::EXCEPTION) {
      throw lvl_;
    }
  }
  static bool verbose;

 private:
  DbgLoggerLvl lvl_;
};
#endif  // OFFLINE_LOGGER_H_
