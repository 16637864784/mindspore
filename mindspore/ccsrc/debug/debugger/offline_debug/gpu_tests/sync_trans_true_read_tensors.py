# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""
Read tensor test script for offline debugger APIs.
"""

import mindspore.offline_debug.dbg_services as d
import numpy as np


def main():

    debugger_backend = d.DbgServices(
        dump_file_path="/home/jtzanaka/dumps/sync_trans_true/032421/alexnet")

    _ = debugger_backend.initialize(
        net_name="Network Name goes here!", is_sync_mode=True)

    # parameter
    info1 = d.TensorInfo(node_name="Default/network-WithLossCell/_backbone-AlexNet/conv2-Conv2d/conv2.bias",
                         slot=0, iteration=2, device_id=0, root_graph_id=0, is_parameter=True)
    # output tensor with zero slot
    info2 = d.TensorInfo(node_name="Default/network-WithLossCell/_backbone-AlexNet/conv3-Conv2d/Conv2D-op308",
                         slot=0, iteration=2, device_id=0, root_graph_id=0, is_parameter=False)
    # output tensor with non-zero slot
    info3 = d.TensorInfo(node_name="Default/network-WithLossCell/_backbone-AlexNet/ReLUV2-op300",
                         slot=1, iteration=2, device_id=0, root_graph_id=0, is_parameter=False)

    tensor_info = [info1, info2, info3]

    tensor_data = debugger_backend.read_tensors(tensor_info)

    print_read_tensors(tensor_info, tensor_data)


def print_read_tensors(tensor_info, tensor_data):
    """Print read tensors."""
    for x, _ in enumerate(tensor_info):
        print("-----------------------------------------------------------")
        print("tensor_info_" + str(x+1) + " attributes:")
        print("node name = ", tensor_info[x].node_name)
        print("slot = ", tensor_info[x].slot)
        print("iteration = ", tensor_info[x].iteration)
        print("device_id = ", tensor_info[x].device_id)
        print("root_graph_id = ", tensor_info[x].root_graph_id)
        print("is_parameter = ", tensor_info[x].is_parameter)
        print()
        print("tensor_data_" + str(x+1) + " attributes:")
        print("data (printed in uint8) = ", np.frombuffer(
            tensor_data[x].data_ptr, np.uint8, tensor_data[x].data_size))
        py_byte_size = len(tensor_data[x].data_ptr)
        c_byte_size = tensor_data[x].data_size
        if c_byte_size != py_byte_size:
            print("The python byte size of ", py_byte_size,
                  " does not match the C++ byte size of ", c_byte_size)
        print("size in bytes = ", tensor_data[x].data_size)
        print("debugger dtype = ", tensor_data[x].dtype)
        print("shape = ", tensor_data[x].shape)


if __name__ == "__main__":
    main()
