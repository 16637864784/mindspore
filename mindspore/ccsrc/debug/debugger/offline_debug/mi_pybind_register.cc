/**
 * Copyright 2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "pybind11/pybind11.h"
#include "pybind11/stl.h"
#include "pybind11/stl_bind.h"
#include "debugger/offline_debug/dbg_services.h"

PYBIND11_MODULE(_mindspore_offline_debug, m) {
  m.doc() = "pybind11 debug services api";
  py::class_<DbgServices>(m, "DbgServices")
    .def(py::init<bool>())
    .def("Initialize", &DbgServices::Initialize)
    .def("AddWatchpoint", &DbgServices::AddWatchpoint)
    .def("RemoveWatchpoint", &DbgServices::RemoveWatchpoint)
    .def("CheckWatchpoints", &DbgServices::CheckWatchpoints)
    .def("ReadTensors", &DbgServices::ReadTensors)
    .def("GetVersion", &DbgServices::GetVersion);

  py::class_<parameter>(m, "parameter")
    .def(py::init<std::string, bool, double, bool, double>())
    .def("get_name", &parameter::get_name)
    .def("get_disabled", &parameter::get_disabled)
    .def("get_value", &parameter::get_value)
    .def("get_hit", &parameter::get_hit)
    .def("get_actual_value", &parameter::get_actual_value);

  py::class_<watchpoint_hit>(m, "watchpoint_hit")
    .def(py::init<std::string, uint32_t, int, uint32_t, std::vector<parameter_t>, int32_t, uint32_t, uint32_t>())
    .def("get_name", &watchpoint_hit::get_name)
    .def("get_slot", &watchpoint_hit::get_slot)
    .def("get_condition", &watchpoint_hit::get_condition)
    .def("get_watchpoint_id", &watchpoint_hit::get_watchpoint_id)
    .def("get_parameters", &watchpoint_hit::get_parameters)
    .def("get_error_code", &watchpoint_hit::get_error_code)
    .def("get_device_id", &watchpoint_hit::get_device_id)
    .def("get_root_graph_id", &watchpoint_hit::get_root_graph_id);

  py::class_<tensor_info>(m, "tensor_info")
    .def(py::init<std::string, uint32_t, uint32_t, uint32_t, uint32_t, bool>())
    .def("get_node_name", &tensor_info::get_node_name)
    .def("get_slot", &tensor_info::get_slot)
    .def("get_iteration", &tensor_info::get_iteration)
    .def("get_device_id", &tensor_info::get_device_id)
    .def("get_root_graph_id", &tensor_info::get_root_graph_id)
    .def("get_is_parameter", &tensor_info::get_is_parameter);

  py::class_<tensor_data>(m, "tensor_data")
    .def(py::init<char *, uint64_t, int, std::vector<int64_t>>())
    .def("get_data_ptr", &tensor_data::get_data_ptr)
    .def("get_data_size", &tensor_data::get_data_size)
    .def("get_dtype", &tensor_data::get_dtype)
    .def("get_shape", &tensor_data::get_shape);
}
