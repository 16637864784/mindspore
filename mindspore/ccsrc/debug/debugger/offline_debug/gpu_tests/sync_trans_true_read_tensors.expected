-----------------------------------------------------------
tensor_info_1 attributes:
node name =  Default/network-WithLossCell/_backbone-AlexNet/conv2-Conv2d/conv2.bias
slot =  0
iteration =  2
device_id =  None
root_graph_id =  0
is_parameter =  True

tensor_data_1 attributes:
data (printed in uint8) =  [  1   0   0   0 195 127   0   0  80  58 118  65 195 127   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0  40 186 117  65
 195 127   0   0   5   0   0   0   0   0   0   0 160  76   6 204 195 127
   0   0  69   0   0   0   0   0   0   0   1   0   0   0 195 127   0   0
  48 135 117  65 195 127   0   0  16  58 118  65 195 127   0   0 144  58
 118  65 195 127   0   0 168 186 117  65 195 127   0   0   6   0   0   0
   0   0   0   0 160  47   6 204 195 127   0   0  69   0   0   0   0   0
   0   0   1   0   0   0 195 127   0   0  80  58 118  65 195 127   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 184 249
 117  65 195 127   0   0   7   0   0   0   0   0   0   0  96  39   6 204
 195 127   0   0  69   0   0   0   0   0   0   0   1   0   0   0 195 127
   0   0 224 218 117  65 195 127   0   0   0   0   0   0   0   0   0   0
 224 219 117  65 195 127   0   0 200  17 118  65 195 127   0   0   8   0
   0   0   0   0   0   0 128 130   5 204 195 127   0   0  69   0   0   0
   0   0   0   0   1   0   0   0 195 127   0   0 120 233 255  59 196 127
   0   0 224 217 117  65 195 127   0   0 224 214 117  65 195 127   0   0
 120 250 117  65 195 127   0   0   9   0   0   0   0   0   0   0 176 213
   4 204 195 127   0   0  69   0   0   0   0   0   0   0   1   0   0   0
 195 127   0   0 240  66 118  65 195 127   0   0 160 218 117  65 195 127
   0   0 224 215 117  65 195 127   0   0  40   9 118  65 195 127   0   0
  10   0   0   0   0   0   0   0  16  46   4 204 195 127   0   0  69   0
   0   0   0   0   0   0   1   0   0   0 195 127   0   0 208  59 118  65
 195 127   0   0   0   0   0   0   0   0   0   0  96 218 117  65 195 127
   0   0  56 251 117  65 195 127   0   0  11   0   0   0   0   0   0   0
  32 137   3 204 195 127   0   0  85   0   0   0   0   0   0   0   1   0
   0   0 195 127   0   0 224 214 117  65 195 127   0   0 144  59 118  65
 195 127   0   0 160 214 117  65 195 127   0   0 136  62 118  65 195 127
   0   0  12   0   0   0   0   0   0   0  48 104  15 204 195 127   0   0
  32 104  15 204 195 127   0   0]
size in bytes =  512
debugger dtype =  11
shape =  [128]
-----------------------------------------------------------
tensor_info_2 attributes:
node name =  Default/network-WithLossCell/_backbone-AlexNet/conv3-Conv2d/Conv2D-op308
slot =  0
iteration =  2
device_id =  None
root_graph_id =  0
is_parameter =  False

tensor_data_2 attributes:
data (printed in uint8) =  [206 239  74 ...  53 201  62]
size in bytes =  4153344
debugger dtype =  11
shape =  [32, 192, 13, 13]
-----------------------------------------------------------
tensor_info_3 attributes:
node name =  Default/network-WithLossCell/_backbone-AlexNet/ReLUV2-op300
slot =  1
iteration =  2
device_id =  None
root_graph_id =  0
is_parameter =  False

tensor_data_3 attributes:
data (printed in uint8) =  [206 239  74 ...  16 239  51]
size in bytes =  831744
debugger dtype =  8
shape =  [207936]
